www/app/app.ts

```diff
@@ -127,13 +125,60 @@ export class App {
   }
 
   showLocalizedError(e?: Error, toastDuration = 10000) {
-    // TODO: Why would these not be set?
-    if (this.rootEl && this.rootEl.async && this.rootEl.showToast) {
-      const msg = getLocalizedErrorMessage(e, this.localize);
-      // Defer this by 500ms so that this toast is shown after any toasts that get shown when any
-      // currently-in-flight domain events land (e.g. fake servers added).
+    let messageKey: string;
+    let messageParams: string[]|undefined;
+    let buttonKey: string;
+    let buttonHandler: () => void;
+    let buttonLink: string;
+
+    if (e instanceof errors.UnexpectedPluginError) {
```

Comment:
I wonder if we can keep the localization separate from the action for better decomposition and only handle the button (key, action, link) for those errors that require it.

Reply:
What are you proposing?

Reply:
I was thinking we could decouple the localization (keeping a function that returns a localized message given an error and message params) from retrieving the button metadata and handlers. Something like:

``` ts
let message = getLocalizedErrorMessage(e, this.localize, ...params);
let buttonKey: string;
let buttonHandler: () => void;
let buttonLink: string;

if (e instanceof errors.ShadowsocksStartFailure) {
  messageKey = 'outline-plugin-error-antivirus';
  buttonKey = 'fix-this';
  buttonLink = 'https://s3.amazonaws.com/outline-vpn/index.html#/en/support/antivirusBlock';
}
...

this.rootEl.showToast(
            message, toastDuration, buttonKey ? this.localize(buttonKey) : undefined, buttonHandler,
            buttonLink);
```

Of course the separation comes at the expense of redundancy, so it's up to you. I think this is fine as you wrote it.

Reply:
Ah, I see what you're saying. I think I'll leave as-is for now, tackle that cleanup some other time.